RUN: only use an overlay for --mount=type=bind,rw on overlay

nalind opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Only use an overlay for --mounts of type=bind,rw, where changes should be discarded, if the storage driver is overlay.  Otherwise, make a temporary copy.

#### How to verify it

Updated CI to test {vfs|overlay} over {vfs|overlay}, which would have caught this.

#### Which issue(s) this PR fixes:

Should fix #5988.

#### Special notes for your reviewer:

Basically what https://github.com/containers/buildah/issues/5988#issuecomment-2718307144 suggested.

#### Does this PR introduce a user-facing change?

```release-note
When not using overlay for storage, running `buildah run` or executing a `buildah build` RUN instruction with the `--mount=type=bind,rw` flag will now use a temporary copy to ensure that writes are discarded, instead of attempting to mount an overlay filesystem using the source location as a "lower".
```

nalind committed:
pkg/overlay: fall back to trying fuse-overlayfs

If we're forcing a mount using overlay, but we fail, and our chosen
upper directory is on overlay, try using fuse-overlayfs.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
RUN: only use an overlay for --mount=type=bind,rw on overlay

Only use an overlay for read-write mounts of type=bind, where changes
should be discarded, if the storage driver is overlay.  Otherwise, make
a temporary copy and use a bind mount to the copy.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
run.bats(run overlay --volume with custom upper and workdir): drop ,z

Drop the conditional ,z that we were trying to add when testing the -v
flag with the "O" option, the combination of which is not allowed by
github.com/containers/common/pkg/parse.ValidateVolumeOpts().

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
imagebuildah.StageExecutor.volumeCacheSaveOverlay(): force mount

Force the mount of the overlay in volumeCacheSaveOverlay(), and clean it
up in volumeCacheRestoreOverlay(), so that we can take advantage of
logic that tries to use fuse-overlayfs if we happen to be on top of
overlay, which would otherwise fail.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
Builder.runSetupVolumeMounts(): force overlay mounts

Force the mount of the overlay in runSetupVolumeMounts() so that we can
take advantage of logic that tries to use fuse-overlayfs if we happen to
be on top of overlay, which would otherwise fail.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
run.bats(root fs only mounted once): also accept rw by itself

When grepping the mounts list for the root mount, also accept the entry
if "rw" is the only flag.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
run.bats(run --mount): skip if we're on overlay

If we're on overlayfs, we can't use the directory name that we'd like to
use as part of the test and expect things to succeed, even if we try to
work around it using fuse-overlayfs.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

siteshwar commented on tests/run.bats:
```diff
@@ -417,32 +411,38 @@ function configure_and_check_user() {
 	run_buildah from --quiet --pull=false $WITH_POLICY_JSON alpine
 	cid=$output
 	mkdir -p ${TEST_SCRATCH_DIR}/was:empty
+	if $(stat -f -c %T ${TEST_SCRATCH_DIR}/chroot/upperdir) = overlayfs; then
```
Please review this finding identified by [OpenScanHub](https://openscanhub.fedoraproject.org/task/53069/log/added.html):

```
Error: SHELLCHECK_WARNING:
/usr/share/buildah/test/system/run.bats:422:58: error[SC2283]: Remove spaces around = to assign (or use [ ] to compare, or quote '=' if literal).
#  420|   	cid=$output
#  421|   	mkdir -p ${TEST_SCRATCH_DIR}/was:empty
#  422|-> 	if $(stat -f -c %T ${TEST_SCRATCH_DIR}/chroot/upperdir) = overlayfs; then
#  423|   		# we'll try to use fuse-overlayfs, which at least through 1.13
#  424|   		# can't accept ":" in layer locations, escaped or not, so bail
```

nalind committed:
overlay.bats(overlay path contains colon): skip if we're on overlay

If we're on overlayfs, we can't use the directory name that we'd like to
use as part of the test and expect things to succeed, even if we try to
work around it using fuse-overlayfs.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

cgwalters commented on internal/volumes/volumes.go:
```diff
@@ -118,6 +120,87 @@ func convertToOverlay(m specs.Mount, store storage.Store, mountLabel, tmpDir str
 	return mountThisInstead, overlayDir, nil
 }
 
+func makeTempCopy(m specs.Mount, tmpDir string, uid, gid int) (specs.Mount, string, error) {
```
This function seems like it could use a doc comment.

cgwalters reviewed (commented):
I only gave this a quick skim and I don't know the code really, but FWIW looks sane to me.

cevich commented on .cirrus.yml:
```diff
@@ -307,7 +307,16 @@ in_podman_task:
         IN_PODMAN: 'true'
         BUILDAH_ISOLATION: 'chroot'
-        STORAGE_DRIVER: 'vfs'
+    matrix:
+        - env:
+              STORAGE_DRIVER: 'vfs'
+        - env:
+              STORAGE_DRIVER: 'overlay'
+    matrix:
+        - env:
+              NESTED_STORAGE_DRIVER: 'vfs'
+        - env:
+              NESTED_STORAGE_DRIVER: 'overlay'
```
I believe cirrus will happily accept this invalid YAML, but I believe `hack/get_ci_vm.sh` will puke on it because it uses python-yaml to parse the file.  If this functionality is important, I believe something like the following would work:

```suggestion
    matrix:
        - env:
              STORAGE_DRIVER: 'vfs'
              NESTED_STORAGE_DRIVER: 'vfs'
        - env:
              STORAGE_DRIVER: 'overlay'
              NESTED_STORAGE_DRIVER: 'vfs'
        - env:
              STORAGE_DRIVER: 'vfs'
              NESTED_STORAGE_DRIVER: 'overlay'
        - env:
              STORAGE_DRIVER: 'overlay'
              NESTED_STORAGE_DRIVER: 'overlay'
```

nalind replied:
Done.

nalind commented on tests/chroot.bats:
```diff
@@ -124,11 +124,24 @@ load helpers
+    if test -x /usr/bin/fuse-overlayfs ; then
+      local storage_opts="overlay.mount_program=/usr/bin/fuse-overlayfs"
+    else
+      skip trying to use overlay on top of overlay, but fuse-overlayfs is not present
```
Done.

TomSweeneyRedHat commented:
@nalind looks lke you need an update here.

Honny1 reviewed (approved):
Thanks, LGTM. I haven't noticed any problems.  I think a rebase will be necessary.

/approve

nalind commented on tests/run.bats:
```diff
@@ -420,23 +420,23 @@ function configure_and_check_user() {
 	mkdir -p ${TEST_SCRATCH_DIR}/was:empty
 	# As a baseline, this should succeed.
-	run_buildah run --mount type=tmpfs,dst=/var/tmpfs-not-empty                                                      $cid touch /var/tmpfs-not-empty/testfile
+	run_buildah run --mount type=tmpfs,dst=/var/tmpfs-not-empty                                                            $cid touch /var/tmpfs-not-empty/testfile
```
Ah, that does help.  Thanks!

mtrmac commented on internal/volumes/volumes.go:
```diff
@@ -329,8 +421,15 @@ func GetBindMount(sys *types.SystemContext, args []string, contextDir string, st
-		if newMount, overlayDir, err = convertToOverlay(newMount, store, mountLabel, tmpDir, 0, 0); err != nil {
-			return newMount, "", "", "", err
+		switch store.GraphDriverName() {
+		case "overlay":
```
(There is precedent for hard-coding a condition like this … but maybe the checks could all be consolidated into a single function.)

mtrmac reviewed (commented):
I don’t know enough to judge the `volumeCache…` part.

Overall, I suppose the added test coverage should be decisive.

TomSweeneyRedHat commented:
@nalind merge conflict here that needs some massaging.

mtrmac commented on contrib/cirrus/lib.sh:
```diff
@@ -198,7 +198,7 @@ in_podman() {
                    "${envargs[@]}" \
                    -e BUILDAH_ISOLATION \
-                   -e STORAGE_DRIVER \
+                   -e STORAGE_DRIVER"${NESTED_STORAGE_DRIVER:+=${NESTED_STORAGE_DRIVER}}" \
```
There used to be changes to `.cirrus.yml` that added `NESTED_STORAGE_DRIVER` to the matrix; I can’t see any current equivalent, was that dropped intentionally?

mtrmac reviewed (commented):
LGTM — although that should carry little weight, I’m very unfamiliar with the problem space.

@nalind is there an ordering dependency with #5975 (in either direction)?

nalind committed:
Run(): handle excludes if we copy the context directory

If we have to copy from the context directory to another location in
order to be able to discard writes to bind-mounts of it, for cases where
we can't just set up an overlay, go ahead and process the excludes for
that context directory, so we at least get a more accurate view of what
should be there for our trouble.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
RUN: handle excludes when overlaying the context directory

If we don't have to copy from the context directory to another location
in order to be able to discard writes to bind-mounts of it, because we
we were able to just set up an overlay over it, go ahead and "apply" the
excludes for that context directory to that mounted overlay.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
add: when adding from a URL, clean up after ourselves

When downloading from a URL source location to add or copy content into
the working container, clean up the temporary directory we created to
hold that content.  We're not reusing it anyway.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
build: clean up downloaded additional build contexts

When we cache the contents of additional build contexts that are
specified by URL, clean them up properly.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>

nalind committed:
Fixup a godoc string

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com>
